mock/endpoints.ts

```diff
@@ -0,0 +1,25 @@
+export default {
```

Comment:
What is the significance of naming this file `mock/endpoints.ts`?  Is the `.ts` something special?

Reply:
Answering both of your points in one, and I'd tried to capture this in my PR description but apparently failed... I found that UMI has a special place in its steely heart reserved for the `mock` directory, and processes any `*.ts` files early as global mocks. So the `endpoints.ts` _is_ the mock for the endpoints API, and it allows `document.ejs` (and the generated `index.html`) to unconditionally `fetch('/api/v1/endpoints')` in both mock and "real" environments; umi's mocking takes care of this transparently.

So this file creates a mock for anyone doing a `GET` on `/api/v1/endpoints`.

This works for me with Gurbir's `start:mock` environment, which pulls up a local dashboard page loaded entirely from local umi/npm dev server mocks. This file has no effect on `build` deployments, and my `fetch('/api/v1/endpoints')` works fine there, too.

The only problem is the e2e environment, which is doing something entirely different. Whether fixing that will require giving up on this neat trick remains to be seen. I'm hoping some of the npm/yarn/umi gurus will have some insight.

Reply:
I think `endpoints.js` works fine here since we don't static-type the object here (which can be an advantage of using a .ts file)

Reply:
Thanks, @aquibbaig; I'm mostly feeling my way through this blindly, as I have very little experience with the npm/yarn/umi development environment. When I discovered the references to the special `.ts` file type, I gained the mistaken impression that was special magic... and since it worked (at least for `start:mock` if not for `test:e2e`), I didn't question that assumption.

You're correct that it works exactly the same as `endpoints.js`. Not surprisingly, though, this also doesn't fix my e2e problems so it doesn't move me forward.

Reply:
@dbutenhof No problem! The e2e tests are fixed as a part of Gurbir's  react-table PR  now.

Reply:
Unfortunately, rebasing on Gurbir's change doesn't solve my e2e failures.

I changed `controllers.e2e.js` to specify `headless: false` and pulled up the debug tab as soon as the window came up. The root problem seems to (still) be that the initial `fetch ( '/api/v1/endpoints' )` is returning the entire `index.html` and apparently not hitting my mock at all.

For `start:mock` the endpoints `fetch` returns the mocked payload; but under the e2e environment it returns

```
<!DOCTYPE html><html lang="en"><head>

<link rel="stylesheet" href="/umi.css">

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title>Pbench Dashboard</title>
  <link rel="icon" href="/favicon.ico" type="image/x-icon">
  
  <!-- runtime endpoints config -->
  <script type="text/javascript">
    fetch( '/api/v1/endpoints' )
      .then(response => {
        console.log(`RESPONSE: ${response.status},${response.statusText}`);
        if (response.ok) {
          response.text().then(data => {
            console.log(`ENDPOINTS: ${data}`);
            window.endpoints = JSON.parse(data);
          }).catch(error => console.log(`ERROR data: ${error}`))
        } else {
          console.log(`NOTOK: ${response}`);
        }
      })
      .catch(error => console.log(`ERROR: ${error}`));
  </script>
<script>
  window.routerBase = "/dashboard/";
</script>
</head>

<body>
  <div id="root"></div>
<script src="/umi.js"></script>



</body></html>
```

And that breaks all of the selectors because the endpoint references like `endpoints.api.controllers_months` don't even resolve.

Reply:
@dbutenhof I just saw that the `controllers.e2e.js` file in your `endpoints` branch is referencing the older file and is different from the current https://github.com/distributed-system-analysis/pbench-dashboard/blob/main/src/e2e/controllers.e2e.js. Also, your endpoint branch looks to be at this state: `This branch is 3 commits ahead, 22 commits behind distributed-system-analysis:main.`

Reply:
Yes, after Gurbir's change went in I merged my local branch and tried things out, but without any change. I didn't bother pushing an update. Just for reference, I'll rebase again and push a new commit to get in sync.

Reply:
![Screenshot](https://user-images.githubusercontent.com/26324376/116280278-82e4f080-a756-11eb-843e-785f109f7622.png)

@dbutenhof  Currently, `windows.endpoints` are getting referenced in the test environment too.

Reply:
I think what you're trying to do is use this endpoint to populate data in the mock environment, is that so?

Reply:
@dbutenhof here's what I found by testing this branch. Idk if it'd be useful or what you intend to do, so do let me know

We used React which listened for changes in the UI and updated the page dynamically, but that's not the case with HTML (unfortunately). That's why we do not see results unless we reload the page on manually script the data into some HTML tag using js. 

- What's happening here is when the mock environment starts and browser fires up, `umi.js` intercepts requests to `/controllers/list` here: https://github.com/distributed-system-analysis/pbench-dashboard/blob/3aae4358bd367b676832821c57f1152a4dfd7c46/src/e2e/controllers.e2e.js#L20.

- Then the dashboard populates the data from the mockfile. After the `umi.js` is called, our script tag is invoked, calling `/api/v1/endpoints`, which fetches the data and updates window.endpoints. But there is no hook to update the changes on the webpage. It just updates the endpoints and stays like that for the rest of the session. Note: umi.js is always invoked at the earliest. Some results from the Network tab.
```
- umi.js 
  -  queued at 48.29ms
  - started at 48.54ms
- endpoints script that we wrote
  - queued at 134.78ms
  - started at 135.40ms 
```
which means that the interception comes earlier than the custom script updating the endpoints, so the data is always fetched from the mock files in the dashboard and not the `/api/v1` endpoint.
 
-  Just manually reloading the browser (without hard reset/cache clean, just a normal reload) can be used to test if it works that way. We can use a `window.location.reload` here but I'd dig deeper into this to find the optimum solution, else there is a race condition between the e2e tests asserting values and fetch API updating the page.